prog/vm/nexus.rb

```diff
@@ -193,18 +199,52 @@ def trigger_refresh_mesh
     hop :run
   end
 
-  def create_ipsec_tunnel(my_subnet, dst_vm, dst_subnet)
-    q_dst_name = dst_vm.inhost_name.shellescape
-    q_dst_net = dst_vm.ephemeral_net6.to_s.shellescape
-
-    my_params = "#{q_vm} #{q_net6} #{my_subnet.to_s.shellescape}"
-    dst_params = "#{q_dst_name} #{q_dst_net} #{dst_subnet.to_s.shellescape}"
+  def create_ipsec_tunnel(my_subnet6, my_subnet4, dst_vm, dst_subnet6, dst_subnet4)
+    src_namespace = vm.inhost_name.shellescape
+    dst_namespace = dst_vm.inhost_name.shellescape
+    src_clover_ephemeral = subdivide_network(vm.ephemeral_net6)
+    dst_clover_ephemeral = subdivide_network(dst_vm.ephemeral_net6)
+    src_private_addr_6 = my_subnet6.to_s.shellescape
+    dst_private_addr_6 = dst_subnet6.to_s.shellescape
+    src_private_addr_4 = my_subnet4.to_s.shellescape
+    dst_private_addr_4 = dst_subnet4.to_s.shellescape
+    src_direction = "out"
+    dst_direction = "fwd"
 
     spi = "0x" + SecureRandom.bytes(4).unpack1("H*")
+    spi4 = "0x" + SecureRandom.bytes(4).unpack1("H*")
     key = "0x" + SecureRandom.bytes(36).unpack1("H*")
 
-    host.sshable.cmd("sudo bin/setup-ipsec setup_src #{my_params} #{dst_params} #{spi} #{key}")
-    dst_vm.vm_host.sshable.cmd("sudo bin/setup-ipsec setup_dst #{my_params} #{dst_params} #{spi} #{key}")
+    # setup source ipsec tunnels
+    host.sshable.cmd("sudo bin/setup-ipsec " \
+      "#{src_namespace} #{src_clover_ephemeral} " \
+      "#{dst_clover_ephemeral} #{src_private_addr_6} " \
+      "#{dst_private_addr_6} #{src_private_addr_4} " \
+      "#{dst_private_addr_4} #{src_direction} " \
+      "#{spi} #{spi4} #{key}")
+
+    # setup destination ipsec tunnels
+    dst_vm.vm_host.sshable.cmd("sudo bin/setup-ipsec " \
+      "#{dst_namespace} #{src_clover_ephemeral} " \
+      "#{dst_clover_ephemeral} #{src_private_addr_6} " \
+      "#{dst_private_addr_6} #{src_private_addr_4} " \
+      "#{dst_private_addr_4} #{dst_direction} " \
+      "#{spi} #{spi4} #{key}")
+  end
+
+  def subdivide_network(net)
+    prefix = net.netmask.prefix_len + 1
+    halved = net.resize(prefix)
+    halved.next_sib
+  end
+
+  def create_private_route(dst_subnet)
+    ### YYY: make this idempotent
+    [dst_subnet.net6, dst_subnet.net4].each do |dst_ip|
+      host.sshable.cmd("sudo ip -n #{q_vm} route add #{dst_ip.to_s.shellescape} dev vethi#{q_vm}")
```

Comment:
```suggestion
    [dst_subnet.net6, dst_subnet.net4].each do |dst_ip|
      host.sshable.cmd("sudo ip -n #{q_vm} route replace #{dst_ip.to_s.shellescape} dev vethi#{q_vm}")
```

I think `replace` does the job? Maybe there are other areas we should be using it, grepping around might be a good idea.

Reply:
I love this!